predicators/explorers/random_nsrts_explorer.py

```diff
@@ -0,0 +1,108 @@
+"""An explorer that takes random NSRTs."""
+
+from typing import Dict, List, Set
+
+from gym.spaces import Box
+
+from predicators import utils
+from predicators.explorers.base_explorer import BaseExplorer
+from predicators.structs import NSRT, Action, DummyOption, \
+    ExplorationStrategy, ParameterizedOption, Predicate, State, Task, Type, \
+    _GroundNSRT, _Option
+
+
+class RandomNSRTsExplorer(BaseExplorer):
+    """RandomNSRTsExplorer implementation.
+
+    Similar to RandomOptionsExplorer in that it chooses
+    uniformly at random out of whichever ground nsrts are
+    applicable in the current state, but different in that
+    the continuous parameter of the parameterized option is
+    generated by the nsrt's sampler rather than by sampling
+    uniformly at random from the option's parameter space.
+
+    This explorer is intended to be used when learning options
+    via reinforcement learning with oracle samplers. In this
+    setting, planning (e.g. using BilevelPlanningExplorer) is
+    not feasible because refinement is slow or fails when the
+    learned options aren't good enough yet.
+
+    Note that the sampler requires the current low-level state
+    as input. An alternative approach would generate a plan of
+    options by planning towards the goal with the ground nsrts
+    and avoid refinement by (1) inferring the terminal low-level
+    state of each option (the initial low-level state of the
+    subsequent option) from the option's sample, which is
+    proposing a low-level subgoal, and (2) setting the simulator
+    state directly to that state. Because applying dimensionality
+    reduction techniques on the sample makes this approach infeasible,
+    and this is something we may want to do, we avoid this approach
+    for now.
+    """
+
+    def __init__(self, predicates: Set[Predicate],
+                 options: Set[ParameterizedOption], types: Set[Type],
+                 action_space: Box, train_tasks: List[Task],
+                 nsrts: Set[NSRT]) -> None:
+
+        super().__init__(predicates, options, types, action_space, train_tasks)
+        self._nsrts = nsrts
+
+        # Sequence of options generated during exploration.
+        self._option_seq: Dict[int, List[_Option]] = {}
```

Comment:
add a comment explaining the key to this dict

Reply:
actually I am not clear on the need for this dict (the random options explorer doesn't need one, for example). if you need it for RL, is it possible to store what's needed inside of the RL approach itself?

Reply:
There are two options:

(1) instantiate an explorer only once (e.g. in `init()` in `nsrt_rl_approach`), which then leads us to creating this dict that maps train task id -> list of options it generated when exploring on that task, because this explorer will keep track of the options its policy generates *for each task*. 

(2) instantiate a new explorer inside `get_interaction_requests()`, so that you have a new explorer for each task. In this case, we don't need the dictionary, but then we'd need to map train task id -> explorer in `nsrt_rl_approach`. 

In `_get_experience_from_result()`, we will ask the explorer what the list of options was for a particular training task. 

I'm fine with either (1) or (2). Do you have a different suggestion for how to do this?

Reply:
For each interaction_result we get back, we want to assign the tuples of (s, a, s') in that to whichever options generated that transition. To do that, we need an *ordered* list of options that we used to generated it. Could you do this from an unordered set of options? Not quite: because two options may be initiable on the same state s, so you wouldn't know which one of the two options generated (s, a, s'). You could figure out by running both options on that state, but that's extra work: since the explorer can easily keep track of which option sequence we used for that interaction result, we might as well use that information. 

So, we might want a map of interaction_request (which leads to an interaction_result) -> list of options used to generate that result. But each interaction_request corresponds to a train_task_idx, so an equivalent mapping is train_task_idx -> list of options. 

Now, *technically*, it would be ok for us to only maintain a SINGLE, *ordered* sequence of options (basically, think of concatenating the list of options from each train_task_idx, and just keep track of which index we are on as we loop through that entire list, when assigning (s, a, s') to an option). But this is more confusing, and prone to more bugs, I think. It's easier to keep track of it separately by train_task_idx (by interaction_result).

Reply:
> we want to assign the tuples of (s, a, s') in that to whichever options generated that transition

Could we accomplish this using a.get_option()? (Note that the option is associated with the action inside of _Option.policy)

Reply:
Oh yeah, I forgot about that, that should work. Will remove the `get_option_sequence()` function, then.